presto-cache/src/main/java/com/facebook/presto/cache/filemerge/FileMergeCacheManager.java

```diff
@@ -235,8 +235,8 @@ public void put(FileReadRequest key, Slice data, CacheQuota cacheQuota)
             return;
         }
 
-        cacheScopeFiles.putIfAbsent(cacheQuota.getIdentifier(), new ConcurrentHashSet<>());
-        cacheScopeFiles.get(cacheQuota.getIdentifier()).add(key.getPath());
+        Set<Path> paths = cacheScopeFiles.computeIfAbsent(cacheQuota.getIdentifier(), k -> new ConcurrentHashSet<>());
```

Comment:
While this change looks reasonable, it looks like there are other race conditions. For example:

```
java.lang.NullPointerException
	at com.facebook.presto.cache.filemerge.FileMergeCacheManager$CacheRemovalListener.onRemoval(FileMergeCacheManager.java:540)
	at com.google.common.cache.LocalCache.processPendingNotifications(LocalCache.java:1809)
	at com.google.common.cache.LocalCache$Segment.runUnlockedCleanup(LocalCache.java:3462)
	at com.google.common.cache.LocalCache$Segment.postWriteCleanup(LocalCache.java:3438)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2145)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3952)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4871)
	at com.facebook.presto.cache.filemerge.FileMergeCacheManager.get(FileMergeCacheManager.java:190)
	at com.facebook.presto.cache.filemerge.TestFileMergeCacheManager.readFully(TestFileMergeCacheManager.java:221)
	at com.facebook.presto.cache.filemerge.TestFileMergeCacheManager.lambda$testStress$0(TestFileMergeCacheManager.java:167)
	at com.facebook.presto.cache.TestingCacheUtils.lambda$stressTest$0(TestingCacheUtils.java:61)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

This error is most likely caused by this code being "racy"

```
            cacheScopeFiles.get(notification.getValue()).remove(path);
            if (cacheScopeFiles.get(notification.getValue()).isEmpty()) {
                cacheScopeFiles.remove(notification.getValue());
            }
```

Also this like looks suspicious. It can in theory result in an NPE.

```
for (Path path : cacheScopeFiles.get(cacheScopeIdentifier))
```

Reply:
I was just looking at the failure. Looks like cacheScopeFiles is very poorly implemented for multi threading use cases. I have salvaged some of the issues, but they are not complete either due to issues in the underlying library. 

But these fixes should eliminate the NPE's .